### PR TITLE
AR model

### PR DESCRIPTION
**Pull request by Jad-yehya**
Add linear AR model implemented in PyTorch. 
Checkpointing, cuda training

### CODE REVIEW

File: benchmark_utils/metrics.py [hunk 1]
```diff
@@ -238,3 +238,24 @@ def ttc(y_true: np.ndarray, y_pred: np.ndarray, return_signed: bool = False):
...
+            Recall score
+
+    Returns
+    -------
+        f1 : float
+            F1 score
+    """
+    if precision + recall == 0:
+        return 0
+    return 2 * (precision * recall) / (precision + recall)
```
Comment (tomMoral) [inline benchmark_utils/metrics.py:261]: Avoid adding unrelated changes in the same PR when you can :)

File: objective.py [hunk 2]
```diff
@@ -48,34 +49,39 @@ def evaluate_result(self, y_hat):
...
         )
+        soft_f1_1 = soft_f1_score(soft_precision1, soft_recall1)
 
         soft_precision3 = soft_precision_score(
             self.y_test, y_hat, detection_range=3
         )
         soft_recall3 = soft_recall_score(
             self.y_test, y_hat, detection_range=3
         )
+        soft_f1_3 = soft_f1_score(soft_precision3, soft_recall3)
```
Comment (tomMoral) [inline objective.py:60]: Do the multiple cases with a loop :)

File: solvers/AR.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        self.model.load_state_dict(best_model)
+
+        self.model.eval()
+        xw_hat = self.model(torch.tensor(
+            self.Xw_test[:, :self.window_size, :],
+            dtype=torch.float32
+        ).to(self.device))
+
+        if self.device == torch.device("cuda"):
+            xw_hat = xw_hat.detach().cpu().numpy()
```
Comment (tomMoral) [inline solvers/AR.py:175]: why not always do this?

File: solvers/AR.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            epoch_loss = 0.0
+
+            for i in range(0, len(self.Xw_train), self.batch_size):
+                x = torch.tensor(
+                    self.Xw_train[i:i+self.batch_size, :self.window_size, :],
+                    dtype=torch.float32
+                )
+                y = torch.tensor(
+                    self.Xw_train[i:i+self.batch_size, :self.horizon, :],
+                    dtype=torch.float32
```
Comment (tomMoral) [inline solvers/AR.py:143]: you can put `device` directly here

File: solvers/AR.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+        "lr": [1e-5],
+        "weight_decay": [1e-7],
+        "window_size": [256],
+        "horizon": [1],
+        "percentile": [99.4],
+    }
+
+    def set_objective(self, X_train, y_test, X_test):
+        self.X_train = X_train
+        self.X_test, self.y_test = X_test, y_test
```
Comment (tomMoral) [inline solvers/AR.py:56]: describe the shape in comments

File: solvers/AR.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+                axis=0
+            ).transpose(0, 2, 1)
+
+        if self.X_test is not None:
+            self.Xw_test = np.lib.stride_tricks.sliding_window_view(
+                X_test,
+                window_shape=self.window_size+self.horizon,
+                axis=0
+            ).transpose(0, 2, 1)
+
```
Comment (tomMoral) [inline solvers/AR.py:84]: is it possible to have `None`?

File: solvers/AR.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        x_hat[self.window_size:self.window_size+self.horizon] = xw_hat[0]
+        # x_hat[self.window_size+self.horizon:] = xw_hat[1:, -self.horizon]
+        x_hat[self.window_size+self.horizon:] = self.mean_overlaping_pred(
+            xw_hat, 1
+        )
+
+        percentile_value = np.percentile(
+            np.abs(self.X_test[self.window_size:] - x_hat[self.window_size:]),
+            self.percentile
+        )
```
Comment (tomMoral) [inline solvers/AR.py:190]: The point-wise error is too volatile. Have you looked at the result?
I would more use the error of the prediciton on the horizon?

Comment (tomMoral) [reply solvers/AR.py:190]: What are the alternative in the paper you read with AR models?

File: tests/test_metrics.py [hunk 8]
```diff
@@ -78,6 +80,34 @@ def test_soft_recall():
                        return_counts=True) == (1.0, 0, 2, 0)
 
 
+def test_soft_f1():
+
+    y1 = np.zeros(10)
+    y1[3] = y1[7] = 1
+
+    precision = soft_precision(y1, y1, detection_range=1)
+    recall = soft_recall(y1, y1, detection_range=1)
```
Comment (tomMoral) [inline tests/test_metrics.py:89]: unused?

**Review by tomMoral — COMMENTED** [review]
Some comments. Nto sure I would have done it with pytorch, why not use `scikit-learn` directly?

Also, could you share the results you get?

File: solvers/AR.py (hunk 6 above)
Comment (Jad-yehya) [reply solvers/AR.py:84]: It is possible to have to X_test or y_test, so for consistency I put for every X_train also

File: tests/test_metrics.py (hunk 8 above)
Comment (Jad-yehya) [reply tests/test_metrics.py:89]: yes indeed, removing it

File: solvers/AR.py (hunk 3 above)
Comment (Jad-yehya) [reply solvers/AR.py:175]: What do you mean ?

File: benchmark_utils/metrics.py (hunk 1 above)
Comment (Jad-yehya) [reply benchmark_utils/metrics.py:261]: Yes will do
